web/old/remctl-2.14/docs/remctl.1
.\" Automatically generated by Pod::Man 2.22 (Pod::Simple 3.07)
.\"
.\" Standard preamble:
.\" ========================================================================
.de Sp \" Vertical space (when we can't use .PP)
.if t .sp .5v
.if n .sp
..
.de Vb \" Begin verbatim text
.ft CW
.nf
.ne \\$1
..
.de Ve \" End verbatim text
.ft R
.fi
..
.\" Set up some character translations and predefined strings.  \*(-- will
.\" give an unbreakable dash, \*(PI will give pi, \*(L" will give a left
.\" double quote, and \*(R" will give a right double quote.  \*(C+ will
.\" give a nicer C++.  Capital omega is used to do unbreakable dashes and
.\" therefore won't be available.  \*(C` and \*(C' expand to `' in nroff,
.\" nothing in troff, for use with C<>.
.tr \(*W-
.ds C+ C\v'-.1v'\h'-1p'\s-2+\h'-1p'+\s0\v'.1v'\h'-1p'
.ie n \{\
.    ds -- \(*W-
.    ds PI pi
.    if (\n(.H=4u)&(1m=24u) .ds -- \(*W\h'-12u'\(*W\h'-12u'-\" diablo 10 pitch
.    if (\n(.H=4u)&(1m=20u) .ds -- \(*W\h'-12u'\(*W\h'-8u'-\"  diablo 12 pitch
.    ds L" ""
.    ds R" ""
.    ds C` ""
.    ds C' ""
'br\}
.el\{\
.    ds -- \|\(em\|
.    ds PI \(*p
.    ds L" ``
.    ds R" ''
'br\}
.\"
.\" Escape single quotes in literal strings from groff's Unicode transform.
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\"
.\" If the F register is turned on, we'll generate index entries on stderr for
.\" titles (.TH), headers (.SH), subsections (.SS), items (.Ip), and index
.\" entries marked with X<> in POD.  Of course, you'll have to process the
.\" output yourself in some meaningful fashion.
.ie \nF \{\
.    de IX
.    tm Index:\\$1\t\\n%\t"\\$2"
..
.    nr % 0
.    rr F
.\}
.el \{\
.    de IX
..
.\}
.\"
.\" Accent mark definitions (@(#)ms.acc 1.5 88/02/08 SMI; from UCB 4.2).
.\" Fear.  Run.  Save yourself.  No user-serviceable parts.
.    \" fudge factors for nroff and troff
.if n \{\
.    ds #H 0
.    ds #V .8m
.    ds #F .3m
.    ds #[ \f1
.    ds #] \fP
.\}
.if t \{\
.    ds #H ((1u-(\\\\n(.fu%2u))*.13m)
.    ds #V .6m
.    ds #F 0
.    ds #[ \&
.    ds #] \&
.\}
.    \" simple accents for nroff and troff
.if n \{\
.    ds ' \&
.    ds ` \&
.    ds ^ \&
.    ds , \&
.    ds ~ ~
.    ds /
.\}
.if t \{\
.    ds ' \\k:\h'-(\\n(.wu*8/10-\*(#H)'\'\h"|\\n:u"
.    ds ` \\k:\h'-(\\n(.wu*8/10-\*(#H)'\`\h'|\\n:u'
.    ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'^\h'|\\n:u'
.    ds , \\k:\h'-(\\n(.wu*8/10)',\h'|\\n:u'
.    ds ~ \\k:\h'-(\\n(.wu-\*(#H-.1m)'~\h'|\\n:u'
.    ds / \\k:\h'-(\\n(.wu*8/10-\*(#H)'\z\(sl\h'|\\n:u'
.\}
.    \" troff and (daisy-wheel) nroff accents
.ds : \\k:\h'-(\\n(.wu*8/10-\*(#H+.1m+\*(#F)'\v'-\*(#V'\z.\h'.2m+\*(#F'.\h'|\\n:u'\v'\*(#V'
.ds 8 \h'\*(#H'\(*b\h'-\*(#H'
.ds o \\k:\h'-(\\n(.wu+\w'\(de'u-\*(#H)/2u'\v'-.3n'\*(#[\z\(de\v'.3n'\h'|\\n:u'\*(#]
.ds d- \h'\*(#H'\(pd\h'-\w'~'u'\v'-.25m'\f2\(hy\fP\v'.25m'\h'-\*(#H'
.ds D- D\\k:\h'-\w'D'u'\v'-.11m'\z\(hy\v'.11m'\h'|\\n:u'
.ds th \*(#[\v'.3m'\s+1I\s-1\v'-.3m'\h'-(\w'I'u*2/3)'\s-1o\s+1\*(#]
.ds Th \*(#[\s+2I\s-2\h'-\w'I'u*3/5'\v'-.3m'o\v'.3m'\*(#]
.ds ae a\h'-(\w'a'u*4/10)'e
.ds Ae A\h'-(\w'A'u*4/10)'E
.    \" corrections for vroff
.if v .ds ~ \\k:\h'-(\\n(.wu*9/10-\*(#H)'\s-2\u~\d\s+2\h'|\\n:u'
.if v .ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'\v'-.4m'^\v'.4m'\h'|\\n:u'
.    \" for low resolution devices (crt and lpr)
.if \n(.H>23 .if \n(.V>19 \
\{\
.    ds : e
.    ds 8 ss
.    ds o a
.    ds d- d\h'-1'\(ga
.    ds D- D\h'-1'\(hy
.    ds th \o'bp'
.    ds Th \o'LP'
.    ds ae ae
.    ds Ae AE
.\}
.rm #[ #] #H #V #F C
.\" ========================================================================
.\"
.IX Title "REMCTL 1"
.TH REMCTL 1 "2009-05-22" "2.14" "remctl"
.\" For nroff, turn off justification.  Always turn off hyphenation; it makes
.\" way too many mistakes in technical documents.
.if n .ad l
.nh
.SH "NAME"
remctl \- Remote execution tool
.SH "SYNOPSIS"
.IX Header "SYNOPSIS"
remctl [\fB\-dhv\fR] [\fB\-p\fR \fIport\fR] [\fB\-s\fR \fIservice\fR] \fIhost\fR \fIcommand\fR
    \fIsubcommand\fR [\fIparameters\fR ...]
.SH "DESCRIPTION"
.IX Header "DESCRIPTION"
\&\fBremctl\fR is a program that allows a user to execute commands remotely on
a server that is running the remctld daemon.  \fBremctl\fR does not interpret
the commands given to it.  It passes them to the server and displays the
return message.  The commands must be defined on the server-side before a
\&\fBremctl\fR client can execute them, and the user running \fBremctl\fR must be
authorized to execute the particular command on the server.
.PP
Access to remote commands is authenticated via Kerberos v5 GSS-API, so a
user must have a ticket granting ticket to use \fBremctl\fR.  All
transmissions to and from the remctld server are encrypted using GSS-API's
security layer.
.PP
\&\fIhost\fR is the hostname of the target server.  \fIcommand\fR and
\&\fIsubcommand\fR together specify the command to run and correspond to the
command names in the configuration file on the server.  \fIparameters\fR are
any additional command-line parameters to pass to the remote command.
.SH "OPTIONS"
.IX Header "OPTIONS"
.IP "\fB\-d\fR" 4
.IX Item "-d"
Turn on extra debugging output of the client-server interaction.
.IP "\fB\-h\fR" 4
.IX Item "-h"
Show a brief usage message and then exit.
.IP "\fB\-p\fR \fIport\fR" 4
.IX Item "-p port"
Connect to the server on \fIport\fR.  If this option isn't given, the client
first tries the registered remctl port (4373) and then falls back on the
legacy port (4444) if that fails.
.IP "\fB\-s\fR \fIservice\fR" 4
.IX Item "-s service"
Authenticate to the server with a service ticket for \fIservice\fR rather
than the default server identity of host/\fIhostname\fR.  This may be
necessary with, for instance, a server where \fBremctld\fR is not running as
root.
.IP "\fB\-v\fR" 4
.IX Item "-v"
Print the version of \fBremctl\fR and exit.
.SH "EXIT STATUS"
.IX Header "EXIT STATUS"
\&\fBremctl\fR will exit with the exit status returned by the remote command.
If some network or authentication error occurred and \fBremctl\fR was unable
to run the remote command or retrieve its exit status, or if \fBremctl\fR was
called with invalid arguments, \fBremctl\fR will exit with status 1.
.SH "EXAMPLES"
.IX Header "EXAMPLES"
Release an \s-1AFS\s0 volume called ls.tripwire:
.PP
.Vb 1
\&    remctl lsdb afs release ls.tripwire
.Ve
.SH "CAVEATS"
.IX Header "CAVEATS"
If no principal is specified with \fB\-s\fR, \fBremctl\fR canonicalizes the
server host name using \s-1DNS\s0 before connecting.  This ensures that the
network connection and the GSS-API authentication use the same server name
even if some common DNS-based load-balancing schemes are in use.  To
disable this canonicalization, specify the server principal using \fB\-s\fR.
.PP
The default behavior, when the port is not specified, of trying 4373 and
falling back to 4444 will be removed in a future version of \fBremctl\fR in
favor of using the \f(CW\*(C`remctl\*(C'\fR service in \fI/etc/services\fR if set and then
falling back on only 4373.  4444 was the poorly-chosen original remctl
port and should be phased out.
.PP
When using Heimdal with triple-DES keys and talking to old servers that
only speak version one of the remctl protocol, \fBremctl\fR may have problems
with \s-1MIC\s0 verification.  This doesn't affect new clients and servers since
the version two protocol doesn't use MICs.  If you are using Heimdal and
run into \s-1MIC\s0 verification problems, see the \s-1COMPATIBILITY\s0 section of
\&\fIgssapi\fR\|(3).
.SH "NOTES"
.IX Header "NOTES"
The remctl port number, 4373, was derived by tracing the diagonals of a
\&\s-1QWERTY\s0 keyboard up from the letters \f(CW\*(C`remc\*(C'\fR to the number row.
.SH "SEE ALSO"
.IX Header "SEE ALSO"
\&\fIkinit\fR\|(1), \fIremctld\fR\|(8)
.PP
The current version of this program is available from its web page at
<http://www.eyrie.org/~eagle/software/remctl/>.
.SH "AUTHOR"
.IX Header "AUTHOR"
Anton Ushakov <antonu@stanford.edu> is the original author.  Updates and
current maintenance are done by Russ Allbery <rra@stanford.edu>.
.SH "COPYRIGHT AND LICENSE"
.IX Header "COPYRIGHT AND LICENSE"
Copyright 2002, 2003, 2004, 2005, 2006, 2007, 2008, 2009 Board of
Trustees, Leland Stanford Jr. University.  All rights reserved.
.PP
Permission to use, copy, modify, and distribute this software and its
documentation for any purpose and without fee is hereby granted, provided
that the above copyright notice appear in all copies and that both that
copyright notice and this permission notice appear in supporting
documentation, and that the name of Stanford University not be used in
advertising or publicity pertaining to distribution of the software
without specific, written prior permission.  Stanford University makes no
representations about the suitability of this software for any purpose.
It is provided \*(L"as is\*(R" without express or implied warranty.
.PP
\&\s-1THIS\s0 \s-1SOFTWARE\s0 \s-1IS\s0 \s-1PROVIDED\s0 \*(L"\s-1AS\s0 \s-1IS\s0\*(R" \s-1AND\s0 \s-1WITHOUT\s0 \s-1ANY\s0 \s-1EXPRESS\s0 \s-1OR\s0 \s-1IMPLIED\s0
\&\s-1WARRANTIES\s0, \s-1INCLUDING\s0, \s-1WITHOUT\s0 \s-1LIMITATION\s0, \s-1THE\s0 \s-1IMPLIED\s0 \s-1WARRANTIES\s0 \s-1OF\s0
\&\s-1MERCHANTABILITY\s0 \s-1AND\s0 \s-1FITNESS\s0 \s-1FOR\s0 A \s-1PARTICULAR\s0 \s-1PURPOSE\s0.


debian/changelog
gutenbach (2.10) unstable; urgency=low

  * Add a useful reminder message to client scripts, "Did you forget to add
    it [the queue] with gutenbach-client-config?"
  * Add 'last' keyword for gbr
  * Use perl bindings for gbr and gbrm in addition to gbq.
  * Update README
  * Add support for pathname expansion/multiple files in gbr
  * Added the "current" option to gbrm, which just removes the current job
    from the queue

 -- Jessica Hamrick <jhamrick@mit.edu>  Tue, 06 Jul 2010 00:11:23 -0400

gutenbach (2.09) unstable; urgency=low

  * Fixed support for playing YouTube videos
  * Fixed error with CUPS_SERVER in gbr and gbrm
  * Added supportfor playing YouTube URLs with gbr

 -- Jessica Hamrick <jhamrick@mit.edu>  Mon, 05 Jul 2010 15:32:18 -0400

gutenbach (2.08) unstable; urgency=low

  * Remove deprecated lpd files in gutenbach-server
  * Modify gutenbach-queue to use /tmp instead of /var/run
  * Replace cupsd.conf with our own version, backing it up to
    cupsd.conf.gutenbach-orig in gutenbach-server
  * Modify gbr and gbrm to use CUPS_SERVER instead of -h in
    gutenbach-client

 -- Jessica Hamrick <jhamrick@mit.edu>  Tue, 29 Jun 2010 19:35:58 -0400

gutenbach (2.07) unstable; urgency=low

  * Update gutenbach-queue to be compatible with CUPS

 -- Jessica Hamrick <jhamrick@mit.edu>  Tue, 29 Jun 2010 01:38:34 -0400

gutenbach (2.06) unstable; urgency=low

  * License gutenbach under GPLv2

 -- Jessica Hamrick <jhamrick@mit.edu>  Mon, 28 Jun 2010 22:46:42 -0400

gutenbach (2.05) unstable; urgency=low

  * Add manpages for gutenbach-client scripts
  * Fix incorrect path in gutenbach-server.install
  * Rename gutenbach to gutenbach-server in po

 -- Jessica Hamrick <jhamrick@mit.edu>  Mon, 28 Jun 2010 22:23:15 -0400

gutenbach (2.04) unstable; urgency=low

  * Create a single source package for all gutenbach-* packages.

 -- Jessica Hamrick <jhamrick@mit.edu>  Sat, 26 Jun 2010 20:53:20 -0400

gutenbach-server (2.03) unstable; urgency=low

  * Update permissions again, to allow everybody to access the server

 -- Jessica Hamrick <jhamrick@mit.edu>  Tue, 08 Jun 2010 19:07:09 -0400

gutenbach-server (2.02) unstable; urgency=low

  * Fix print server permissions problems such that remote
  users can user cups-lpq and cups-lprm

 -- Jessica Hamrick <jhamrick@mit.edu>  Mon, 31 May 2010 23:51:52 -0400

gutenbach-server (2.01) unstable; urgency=low

  * Handle SIGTERM in the gutenbach backend, so jobs don't
  keep playing even after being dequeued.

 -- Jessica Hamrick <jhamrick@mit.edu>  Mon, 31 May 2010 22:23:53 -0400

gutenbach-server (2.0) unstable; urgency=low

  * Changed from using LPRng to CUPS

 -- Jessica Hamrick <jhamrick@mit.edu>  Thu, 13 May 2010 17:40:53 -0400

gutenbach-server (1.01) unstable; urgency=low

  * Implemented debconf to retrieve printer and zephyr class names.
  * Made uninstall cleaner (and when you purge, it purges
  configuration files and changes)

 -- Jessica Hamrick <jhamrick@mit.edu>  Wed, 16 Sep 2009 15:58:40 -0400

gutenbach-server (1.0) unstable; urgency=low

  * Initial release.

 -- Jessica Hamrick <jhamrick@mit.edu>  Tue, 08 Sep 2009 18:38:12 -0400

gutenbach-client (1.01-1) unstable; urgency=low

  * Make package no longer Debian native.

 -- Jessica Hamrick <jhamrick@mit.edu>  Sat, 26 Jun 2010 19:20:08 -0400

gutenbach-client (1.01) unstable; urgency=low

  * Fix debhelper compatibility to 5 (for backwards compatibility)

 -- Jessica Hamrick <jhamrick@mit.edu>  Sat, 26 Jun 2010 18:01:02 -0400

gutenbach-client (1.0) unstable; urgency=low

  * Initial package build
  * gbq - show queue
  * gbr - queue file
  * gbrm - remove file
  * gutenbach-client-config - add, list, or delete queues

 -- Jessica Hamrick <jhamrick@mit.edu>  Sat, 26 Jun 2010 16:56:28 -0400

gutenbach-client (0.01) unstable; urgency=low

  * Created skeleton package.

 -- Jessica Hamrick <jhamrick@mit.edu>  Sat, 19 Jun 2010 00:23:53 -0400

gutenbach-queue (1.02) unstable; urgency=low

  * Prevent the screen from turning off when
  view-gutenbach-queue is running.

 -- Jessica Hamrick <jhamrick@mit.edu>  Mon, 14 Dec 2009 20:24:37 -0500

gutenbach-queue (1.01) unstable; urgency=low

  * Changed the chown in the postinst to
  daemon:daemon instaed of daemon:pika

 -- Jessica Hamrick <jhamrick@mit.edu>  Mon, 14 Dec 2009 20:23:50 -0500

gutenbach-queue (1.0) unstable; urgency=low

  * Initial release.

 -- Jessica Hamrick <jhamrick@mit.edu>  Mon, 14 Dec 2009 20:03:48 -0500

gutenbach-remctl (1.02) unstable; urgency=low

  * Depend on version 2.0 or greater of gutenbach

 -- Jessica Hamrick <jhamrick@mit.edu>  Thu, 13 May 2010 17:42:53 -0400

gutenbach-remctl (1.01) unstable; urgency=low

  * Just say the new volume, not who changed it.

 -- Jessica Hamrick <jhamrick@mit.edu>  Mon, 14 Dec 2009 22:04:43 -0500

gutenbach-remctl (1.0) unstable; urgency=low

  * Initial release.

 -- Jessica Hamrick <jhamrick@mit.edu>  Mon, 14 Dec 2009 22:03:44 -0500


web/old/remctl-2.14/client/remctl.c
/*
 * remctl command-line client.
 *
 * This is a command-line driver for the libremctl library, which takes the
 * command on the command line and prints out the results to standard output
 * and standard error as appropriate.
 *
 * Originally written by Anton Ushakov
 * Extensive modifications by Russ Allbery <rra@stanford.edu>
 * Copyright 2002, 2003, 2004, 2005, 2006, 2007, 2008, 2009
 *     Board of Trustees, Leland Stanford Jr. University
 *
 * See LICENSE for licensing terms.
*/

#include <config.h>
#include <portable/system.h>
#include <portable/getopt.h>
#include <portable/socket.h>

#include <ctype.h>

#include <client/remctl.h>
#include <util/util.h>

/* Usage message. */
static const char usage_message[] = "\
Usage: remctl <options> <host> <command> <subcommand> <parameters>\n\
\n\
Options:\n\
    -d            Debugging level of output\n\
    -h            Display this help\n\
    -p <port>     remctld port (default: 4373 falling back to 4444)\n\
    -s <service>  remctld service principal (default: host/<host>)\n\
    -v            Display the version of remctl\n";


/*
 * Display the usage message for remctl.
 */
static void
usage(int status)
{
    fprintf((status == 0) ? stdout : stderr, "%s", usage_message);
    exit(status);
}


/*
 * Get the responses back from the server, taking appropriate action on each
 * one depending on its type.  Sets the errorcode parameter to the exit status
 * of the remote command, or to 1 if the remote command failed with an error.
 * Returns true on success, false if some protocol-level error occurred when
 * reading the responses.
 */
static bool
process_response(struct remctl *r, int *errorcode)
{
    struct remctl_output *out;

    *errorcode = 0;
    out = remctl_output(r);
    while (out != NULL && out->type != REMCTL_OUT_DONE) {
        switch (out->type) {
        case REMCTL_OUT_OUTPUT:
            if (out->stream == 1)
                fwrite(out->data, out->length, 1, stdout);
            else if (out->stream == 2)
                fwrite(out->data, out->length, 1, stderr);
            else {
                warn("unknown output stream %d", out->stream);
                fwrite(out->data, out->length, 1, stderr);
            }
            break;
        case REMCTL_OUT_ERROR:
            *errorcode = 255;
            fwrite(out->data, out->length, 1, stderr);
            fputc('\n', stderr);
            return true;
        case REMCTL_OUT_STATUS:
            *errorcode = out->status;
            return true;
        case REMCTL_OUT_DONE:
            break;
        }
        out = remctl_output(r);
    }
    if (out == NULL) {
        die("error reading from server: %s", remctl_error(r));
        return false;
    } else
        return true;
}


/*
 * Main routine.  Parse the arguments, open the remctl connection, send the
 * command, and then call process_response.
 */
int
main(int argc, char *argv[])
{
    int option, status;
    char *server_host;
    struct addrinfo hints, *ai;
    char *service_name = NULL;
    unsigned short port = 0;
    struct remctl *r;
    int errorcode = 0;

    /* Set up logging and identity. */
    message_program_name = "remctl";
    if (!socket_init())
        die("failed to initialize socket library");

    /*
     * Parse options.  The + tells GNU getopt to stop option parsing at the
     * first non-argument rather than proceeding on to find options anywhere.
     * Without this, it's hard to call remote programs that take options.
     * Non-GNU getopt will treat the + as a supported option, which is handled
     * below.
     */
    while ((option = getopt(argc, argv, "+dhp:s:v")) != EOF) {
        switch (option) {
        case 'd':
            message_handlers_debug(1, message_log_stderr);
            break;
        case 'h':
            usage(0);
            break;
        case 'p':
            port = atoi(optarg);
            break;
        case 's':
            service_name = optarg;
            break;
        case 'v':
            printf("%s\n", PACKAGE_STRING);
            exit(0);
            break;
        case '+':
            fprintf(stderr, "%s: invalid option -- +\n", argv[0]);
        default:
            usage(1);
            break;
        }
    }
    argc -= optind;
    argv += optind;
    if (argc < 3)
        usage(1);
    server_host = *argv++;
    argc--;

    /*
     * If service_name isn't set, the remctl library uses host/<server>
     * (host@<server> in GSS-API parlance).  However, if the server to which
     * we're connecting is a DNS-load-balanced name, we have to be careful
     * what principal name we use.
     *
     * Ideally, we would let the GSS-API library handle this and choose
     * whether to canonicalize the <server> in the principal name based on the
     * krb5.conf rdns setting and similar configuration.  However, with DNS
     * load balancing, this still may fail.  At the time of network
     * connection, we will connect to whatever the name resolves to then.
     * After we connect, we authenticate, and the GSS-API library will then
     * separately canonicalize the hostname.  It could get a different answer
     * than we got for our network connection, leading to an authentication
     * failure.
     *
     * Therefore, if the principal isn't specified, we canonicalize the
     * hostname to which we're connecting before we connect.  Then, the
     * additional canonicalization possibly done by the GSS-API library should
     * return the same results and be consistent.
     *
     * Note that this opens the possibility of a subtle attack through DNS
     * spoofing, since both the principal used and the host to which we're
     * connecting can be changed by varying the DNS response.
     *
     * If the principal is specified explicitly, assume the user knows what
     * they're doing and don't do any of this.
     */
    if (service_name == NULL) {
        memset(&hints, 0, sizeof(hints));
        hints.ai_flags = AI_CANONNAME;
        status = getaddrinfo(server_host, NULL, &hints, &ai);
        if (status != 0)
            die("cannot resolve host %s: %s", server_host,
                gai_strerror(status));
        server_host = xstrdup(ai->ai_canonname);
        freeaddrinfo(ai);
    }

    /* Open connection. */
    r = remctl_new();
    if (r == NULL)
        sysdie("cannot initialize remctl connection");
    if (!remctl_open(r, server_host, port, service_name))
        die("%s", remctl_error(r));

    /* Do the work. */
    if (!remctl_command(r, (const char **) argv))
        die("%s", remctl_error(r));
    if (!process_response(r, &errorcode))
        die("%s", remctl_error(r));

    /* Shut down cleanly. */
    remctl_close(r);
    socket_shutdown();
    return errorcode;
}
